tensorflow/core/kernels/conv_ops_gpu_3.cu.cc

```diff
@@ -413,6 +429,275 @@ struct PadInput<GPUDevice, T, int, NDIMS> {
+    else if (tile_size_j <= TileLongSide && tile_size_i <= TileShortSide)
+      SwapDimension1And2InTensor3UsingTiles<
+          T, NumThread, TileShortSide,
+          TileLongSide><<<total_tiles_count, NumThread, 0, d.stream()>>>(
+          input, input_dims, output);
+    return;
+  }
+};
+
+template <typename T, int TileLongSide, int TileShortSide>
+struct BatchNarrowMatrixTransposeDispatcher<
+    T, TileLongSide, TileShortSide,
+    typename std::enable_if<TileSizePossibilityFrontierCheck(
+                                TileLongSide, TileShortSide, sizeof(T),
+                                /* mode = outside */ 1),
+                            void>::type> {
+  static void DoIt(const GPUDevice& d, int tile_size_i,
+                                           int tile_size_j,
+                                           int total_tiles_count,
+                                           const T* input,
+                                           const Dimension<3>& input_dims,
+                                           T* output) {
+    assert(false &&
+           "BatchNarrowMatrixTransposeDispatcher has requested an unexpected "
+           "launch configuration. ");
+  }
+};
+
+// A helper function to make RunSwapDimension1And2InTensor3 concise. This
+// helper function looks at the data type and input matrix sizes and decides
+// the thread numbers and tile sizes to use.
+template <typename T>
+void SwapDimension1And2InTensor3WithNarrowMatrices(
+    const GPUDevice& d, const T* input, const Dimension<3>& input_dims,
+    T* output, const int kMinDimensionToUseTiles) {
+  // Define available tile sizes here for each size of data type supported:
+  std::vector<std::pair<int, int>> tile_spec =
+      []() -> std::vector<std::pair<int, int>> {
+    switch (sizeof(T)) {
+      case 1:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 2:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 4:
+        return {{32, 15}, {64, 15}, {128, 15}, {256, 15}, {512, 4}, {1024, 2}};
+      case 8:
+        return {{32, 15}, {64, 15}, {128, 8}, {256, 4}, {512, 2}};
+      case 16:
+        return {{32, 4},  {64, 4},  {128, 4}, {256, 2}};
+    }
+    return {};
+  }();
+
+  int tile_long_side_len = 0;
+  int tile_short_side_len = 0;
+  float lowest_cost = std::numeric_limits<float>::max();
+  int data_long_side = max(input_dims[1], input_dims[2]);
+
+  for (auto tile_size_pair : tile_spec) {
+    int proposed_tile_long_side_len = tile_size_pair.first;
+
+    // Threads that will not be doing anything useful when reading the matrix
+    // because the thread block size is bigger than the data block size.
+    int wasted_threads = (data_long_side -
+                          MathUtil::FloorOfRatio<int>(
+                              data_long_side, proposed_tile_long_side_len) *
+                              proposed_tile_long_side_len);
```

Comment:
I know it's picky, but I think it would be great to clang-format these commits.  For example, I'm pretty sure clang-format will pull the last line of this statement in by 4 spaces, which makes the computation much easier to follow.  Using a relatively new version of clang-format will also format the `<<<foo>>>` kernel launch nicely.

`git-clang-format` will let you format the commits without reformatting the whole file.